src/libraries/System.Text.Json/tests/Serialization/ExceptionTests.cs

```diff
@@ -527,4 +527,100 @@ public static void UnsupportedTypeFromRoot()
             Assert.DoesNotContain("Path: $", ex.Message);
         }
     }
+
+    public class OpenGenericTests_Span : OpenGenericTests
+    {
+        public OpenGenericTests_Span() : base(SerializationWrapper.SpanSerializer) { }
+    }
+
+    public class OpenGenericTests_String : OpenGenericTests
+    {
+        public OpenGenericTests_String() : base(SerializationWrapper.StringSerializer) { }
+    }
+
+    public class OpenGenericTests_Stream : OpenGenericTests
+    {
+        public OpenGenericTests_Stream() : base(SerializationWrapper.StreamSerializer) { }
+    }
+
+    public class OpenGenericTests_StreamWithSmallBuffer : OpenGenericTests
+    {
+        public OpenGenericTests_StreamWithSmallBuffer() : base(SerializationWrapper.StreamSerializerWithSmallBuffer) { }
+    }
+
+    public class OpenGenericTests_Writer : OpenGenericTests
+    {
+        public OpenGenericTests_Writer() : base(SerializationWrapper.WriterSerializer) { }
+    }
+
+    public abstract class OpenGenericTests
+    {
+        private SerializationWrapper Serializer { get; }
+
+        public OpenGenericTests(SerializationWrapper serializer)
+        {
+            Serializer = serializer;
+        }
+
+        [Fact]
+        public void DeserializeOpenGeneric()
+        {
+            foreach (object[] obj in TypesWithOpenGenerics_ToDeserialize())
+            {
+                Type type = (Type)obj[0];
+
+                InvalidOperationException ex = Assert.Throws<InvalidOperationException>(() => JsonSerializer.Deserialize("", type));
+                Assert.Contains(type.ToString(), ex.ToString());
+            }
+        }
+
+        [Fact]
+        public void SerializeOpenGeneric()
+        {
+            object obj;
+
+            foreach (object[] types in TypesWithOpenGenerics_ToSerialize())
```

Comment:
nit: use `MemberData` and change this to `Theory` and then you can remove the `foreach`.

Reply:
Xunit does not support `Theory` tests when using non-`static` test classes. See https://github.com/dotnet/runtime/pull/33815 for a related discussion.

Reply:
Do you have a repro? I was able to switch from `Fact` to `Theory` without any issue and it looks to me that all the tests on the derived types ran correctly.
```
dotnet build /t:test /p:xunitmethodname=*SerializeOpenGeneric
```
```cs
[Theory]
[MemberData(nameof(TypesWithOpenGenerics_ToSerialize))]
public void SerializeOpenGeneric(Type typex)
{
    object obj;

    InvalidOperationException ex;

    //foreach (object[] types in TypesWithOpenGenerics_ToSerialize())
    //{
    Type type = typex;//(Type)types[0];

        if (type.GetGenericArguments().Length == 1)
        {
            obj = Activator.CreateInstance(type.MakeGenericType(typeof(int)));
        }
        else
        {
            obj = Activator.CreateInstance(type.MakeGenericType(typeof(string), typeof(int)));
        }

        Assert.Throws<ArgumentException>(() => Serializer.Serialize(obj, type));

        ex = Assert.Throws<InvalidOperationException>(() => Serializer.Serialize(null, type));
        Assert.Contains(type.ToString(), ex.ToString());
    //}

    Type openNullableType = typeof(Nullable<>);

    obj = Activator.CreateInstance(openNullableType.MakeGenericType(typeof(int)));
    ex = Assert.Throws<InvalidOperationException>(() => Serializer.Serialize(obj, openNullableType));
    Assert.Contains(openNullableType.ToString(), ex.ToString());
}
```

```
c:\repos\runtime\src\libraries\System.Text.Json\tests>dotnet build /t:test /p:xunitmethodname=*SerializeOpenGeneric
Microsoft (R) Build Engine version 16.6.0-preview-20112-08+77da97f69 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  Restore completed in 50.34 ms for c:\repos\runtime\src\libraries\Common\tests\CoreFx.Private.TestUtilities\CoreFx.Private.TestUtilities.csproj.
  Restore completed in 50.34 ms for c:\repos\runtime\src\libraries\System.Text.Json\tests\System.Text.Json.Tests.csproj.
  CoreFx.Private.TestUtilities -> c:\repos\runtime\artifacts\bin\CoreFx.Private.TestUtilities\netcoreapp5.0-Debug\CoreFx.Private.TestUtilities.dll
  System.Text.Json.Tests -> c:\repos\runtime\artifacts\bin\System.Text.Json.Tests\netcoreapp5.0-Debug\System.Text.Json.Tests.dll
  ----- start Mon 03/30/2020 15:55:08.58 ===============  To repro directly: =====================================================
  pushd c:\repos\runtime\artifacts\bin\System.Text.Json.Tests\netcoreapp5.0-Debug\
  "c:\repos\runtime\artifacts\bin\testhost\netcoreapp5.0-Windows_NT-Debug-x64\dotnet.exe" exec --runtimeconfig System.Text.Json.Tests.runtimeconfig.json --depsfile System.Text.Json.Tests.deps.json xunit.console.dll System.Text.Json.Tests.dll -xml testResults.xml -nologo -method *SerializeOpenGeneric -notrait category=OuterLoop -notrait category=failing -notrait category=nonnetcoreapptests -notrait category=nonwindowstests
  popd
  ===========================================================================================================
    Discovering: System.Text.Json.Tests (method display = ClassAndMethod, method display options = None)
    Discovered:  System.Text.Json.Tests (found 5 of 1779 test cases)
    Starting:    System.Text.Json.Tests (parallel test collections = on, max threads = 12)
    Finished:    System.Text.Json.Tests
  === TEST EXECUTION SUMMARY ===
     System.Text.Json.Tests  Total: 15, Errors: 0, Failed: 0, Skipped: 0, Time: 0.253s
  ----- end Mon 03/30/2020 15:55:09.78 ----- exit code 0 ----------------------------------------------------------

Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:04.91
```

Reply:
Hmm weird. It's running all the tests now, as you've shown. This is different from my earlier experiences described in https://github.com/dotnet/runtime/pull/33815, and when I was writing tests for https://github.com/dotnet/runtime/pull/33444 and this PR. Not sure if there was a change to Xunit/our test infra.

I'll look out for this behavior, and I'll use `Theory` here.

Reply:
I believe the issue is that you can't specify a base class for xunit and expect all derived tests to run.